feat(git)!: determine branch modification based on all branch commits

AaronMoat opened the pull request:
## Changes

Extract all authors from the git log of the changes in the branch for a given PR against master and use those in isBranchModified

## Context

Currently, the documentation for `gitIgnoredAuthors` states:

> Specify commit authors ignored by Renovate.

> By default, Renovate will treat any PR as modified if another Git author has added to the branch. When a PR is considered modified, Renovate won't perform any further commits such as if it's conflicted or needs a version update. If you have other bots which commit on top of Renovate PRs, and don't want Renovate to treat these PRs as modified, then add the other Git author(s) to gitIgnoredAuthors.

The current implementation, however, only looks at the latest commit. If you have a flow like:

1. Renovate commit
2. Human commit
3. Bot commit (e.g. formatting) 

This will mean isBranchModified returns false and the rebase flow throws away commits 2 and 3. 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=28225) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=28225) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=28225) it.</sub>

samchungy commented on lib/util/git/index.ts:
```diff
@@ -651,18 +654,13 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
-      ])
-    ).trim();
+    const commits = baseBranch
+      ? await git.log([`origin/main..origin/${branchName}`])
```
Or does base branch include origin?

AaronMoat replied:
i dont think it does

rarkins reviewed (changes requested):
We had a couple of attempts at this same concept and both times it caused unintended consequences causing it to be rolled back. It's important we understand why that happened and how this fresh attempt avoids the same problem before merging

AaronMoat commented:
Good call @rarkins I hadn't thought about checking closed issues/discussions so I hadn't seen that context.

I have found https://github.com/renovatebot/renovate/pull/21505/ - this comment stands out to me https://github.com/renovatebot/renovate/issues/21504#issuecomment-1507972161 - and I support the assessment because the attempt used `origin/${branchName}...origin/${baseBranch}`. This is wrong: it includes all commits that are different between the two branches, not just the new ones. https://stackoverflow.com/a/24186641/16823396 

Are you able to find what the other issue was?

AaronMoat commented:
Ah, was this the second time? https://github.com/renovatebot/renovate/pull/21558/files#r1361930608 

I think this attempt also fixed that problem. 

That being said, if it's been wrong twice, who's to say it can't be a third time. What do you recommend to gain confidence?

RahulGautamSingh commented:
> That being said, if it's been wrong twice, who's to say it can't be a third time. What do you recommend to gain confidence?

This PR does things correctly I think ie. instead of the wrong way `origin/${branchName}...origin/${baseBranch}` it uses the correct one `origin/${baseBranch}...origin/${branchName}`.

In both the previous attempts the wrong way was merged. I think some real runs will help us gain confidence.
If you can first test all the cases which made the previous attempts fail and also test normal cases. Then we can discuss from there.
Some example test runs:
1. simple update PR (up-to-date with base branch and no other commits except bot's)
2. simple update PR (behind base branch and no other commits except bot's)
3. modified update PR 
4. conflicted update PR 
5. rebased update PR but no external commits
6. rebased updated PR but with external commits (ie. a modified PR which has been rebased after user requested)

Some other mixes like simple update PR with commits from ignored authors.

AaronMoat commented:
This is the test cases without using the gitIgnoredAuthors configuration. 

1. ✅ Up to date, clean PR happily ignored ([1.run-on-update-pr.log](https://github.com/renovatebot/renovate/files/14892402/1.run-on-update-pr.log)) https://github.com/AaronMoat/renovate-testing/pull/2/commits/85e6cfeb699ae57f1935663d01dd8a2a1a0640a7 
2. ✅ PR that's behind is happy ignored ([2.run-on-update-pr-when-behind.log](https://github.com/renovatebot/renovate/files/14892404/2.run-on-update-pr-when-behind.log)) or rebased with `behind-base-branch` ([2.behind-base-with-rebasewhen-behind-base-branch.log](https://github.com/renovatebot/renovate/files/14892403/2.behind-base-with-rebasewhen-behind-base-branch.log) - https://github.com/AaronMoat/renovate-testing/commit/d95d200eaefa4b57fe0d0250ba108d9306999223)
3. ✅ PR that's been modified successfully gets the "Edited/Blocked Notification". https://github.com/AaronMoat/renovate-testing/pull/2#issuecomment-2040949983 ([3.modified-pr.log](https://github.com/renovatebot/renovate/files/14892405/3.modified-pr.log))
4. ✅ rebased ([4.conflicted.log](https://github.com/renovatebot/renovate/files/14892406/4.conflicted.log), https://github.com/AaronMoat/renovate-testing/commit/460a58f68cef8c225601c544ac1852e4168d13ac)
5. Sorry, I'm a bit confused. What does rebased update PR mean? 
6. ^
7. ✅ It shows as modified: https://github.com/AaronMoat/renovate-testing/pull/3#issuecomment-2040966080 ([7.github-ui-update.log](https://github.com/renovatebot/renovate/files/14892407/7.github-ui-update.log))

I will do some testing with the gitIgnoredAuthors configuration once we're satisfied that this hasn't broken any of the standard behaviour.

viceice commented on lib/util/git/index.ts:
```diff
@@ -628,7 +628,10 @@ export async function isBranchBehindBase(
-export async function isBranchModified(branchName: string): Promise<boolean> {
+export async function isBranchModified(
+  branchName: string,
+  baseBranch?: string,
```
is there a use case where we don't pass the base branch?

AaronMoat replied:
Good question! I meant to call that out in the PR description and forgot. 

I've pushed a commit that might clarify it -- I wasn't able to immediately come up with a base branch for the onboarding and migration flows that made a lot of sense, given I don't know much about them, but happy to workshop them and perhaps if there is a sensible choice we could make it required.

RahulGautamSingh commented:
> Sorry, I'm a bit confused. What does rebased update PR mean?

In update PR created by the bot. There is a checkbox present which user can check to rebase the PR.
```  [] If you want to rebase/retry this PR, check this box ```

When you click on it the bot [rebases ](https://docs.renovatebot.com/updating-rebasing/)the PR.

AaronMoat commented:
Cool, rebase is fine too (updated prior comment). 

As well,

- https://github.com/AaronMoat/renovate-testing/pull/7 -> committed with a bot author and ran, got isModified = false (and no "blocked" message), [gitIgnoredAuthor-latest.log](https://github.com/renovatebot/renovate/files/14895833/gitIgnoredAuthor-latest.log)
- https://github.com/AaronMoat/renovate-testing/pull/5 -> committed with a human then a bot, got isModified = true (and a "blocked" message), 
[gitIgnoredAuthor-with-human-then-bot.log](https://github.com/renovatebot/renovate/files/14895836/gitIgnoredAuthor-with-human-then-bot.log)

This all seems to be working as expected for the cases I have tested.

viceice commented on lib/modules/platform/local/scm.ts:
```diff
@@ -9,7 +9,7 @@ export class LocalFs implements PlatformScm {
     return Promise.resolve(false);
   }
-  isBranchModified(branchName: string): Promise<boolean> {
+  isBranchModified(branchName: string, baseBranch: string): Promise<boolean> {
```
```suggestion
  isBranchModified(branchName: string, baseBranch: string | undefined): Promise<boolean> {
```

follow interface declaration

viceice commented on lib/util/git/index.ts:
```diff
@@ -628,7 +628,10 @@ export async function isBranchBehindBase(
-export async function isBranchModified(branchName: string): Promise<boolean> {
+export async function isBranchModified(
+  branchName: string,
+  baseBranch: string | undefined,
```
i think when `baseBranch` is undefined then we need to use `config.defaultBranch` 🤔

AaronMoat replied:
Is your recommendation to use config.defaultBranch for the onboarding and migration flows and make this required rather than optional?

rarkins replied:
Yes, I'd prefer that

viceice commented on lib/modules/platform/types.ts:
```diff
@@ -270,7 +270,10 @@ export interface Platform {
-  isBranchModified(branchName: string): Promise<boolean>;
+  isBranchModified(
+    branchName: string,
+    baseBranch: string | undefined,
```
```suggestion
    baseBranch?: string | undefined,
```

it should probably stay optional to have less code changes. so we don't need to explicit pass undefined

rarkins replied:
In which case have we no baseBranch though? I would lean towards mandatory if there's no use case where we need told behavior

rarkins reviewed (changes requested):
Make baseBranch mandatory

viceice commented on lib/util/git/index.ts:
```diff
@@ -651,18 +654,13 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+      `origin/${baseBranch}..origin/${branchName}`,
+    ]);
+
+    commits.all.forEach((commit) => committedAuthors.add(commit.author_email));
```
use for-of loop

viceice commented on lib/workers/repository/config-migration/branch/rebase.ts:
```diff
@@ -15,8 +15,9 @@ export async function rebaseMigrationBranch(
   migratedConfigData: MigratedData,
 ): Promise<string | null> {
   logger.debug('Checking if migration branch needs rebasing');
+  const baseBranch = config.defaultBranch!;
```
@rarkins should we use `baseBranch` instead of `defaultBranch` here? do we support config migration on multiple base branches? 🤔

rarkins replied:
Do we migrate configs in base branches? It's possible we only migrate in the default branch so this is correct

AaronMoat replied:
Based on my reading of https://github.com/renovatebot/renovate/blob/4129ab2077a28c551bf8d48d8d80388a8a741289/lib/workers/repository/config-migration/pr/index.ts#L101, I understand that it's only the default branch.

rarkins replied:
Yes I think it's only default branch, so a hardcoded default branch here is ok

viceice commented on lib/modules/platform/types.ts:
```diff
@@ -270,7 +270,7 @@ export interface Platform {
   isBranchBehindBase(branchName: string, baseBranch: string): Promise<boolean>;
-  isBranchModified(branchName: string): Promise<boolean>;
+  isBranchModified(branchName: string, baseBranch: string): Promise<boolean>;
   isBranchConflicted(baseBranch: string, branch: string): Promise<boolean>;
```
we should flip this in another pr to align with other functions above

AaronMoat replied:
Agree, this one was bugging me :)

viceice commented on lib/util/git/index.ts:
```diff
@@ -651,18 +654,15 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
   await syncGit();
-  // Retrieve the author of the most recent commit
-  let lastAuthor: string | undefined;
+  const committedAuthors: Set<string> = new Set();
```
```suggestion
  const committedAuthors = new Set<string>();
```
simplify 😉

viceice commented on lib/util/git/index.ts:
```diff
@@ -673,19 +673,44 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+    includedAuthors.delete(gitAuthorEmail);
+  }
+
+  ignoredAuthors?.forEach((ignoredAuthor) => {
+    includedAuthors.delete(ignoredAuthor);
+  });
```
```suggestion
  for (const ignoredAuthor of ignoredAuthors ?? []) {
    includedAuthors.delete(ignoredAuthor);
  }
```

rarkins commented:
Should we merge this in v38 just in case it has any breaking changes/implications?

AaronMoat commented:
> 
Should we merge this in v38 just in case it has any breaking changes/implications?

Personally fine with this, but not really up to me! Do you mean to mark it as a breaking change `fix(git)!`?

rarkins commented:
I've retargetted it to v38

viceice commented on lib/util/git/index.ts:
```diff
@@ -673,19 +673,44 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+    includedAuthors.delete(gitAuthorEmail);
+  }
+
+  for (const ignoredAuthor of ignoredAuthors) {
```
are we sure ignoredAuthors is never null or undefined? 🤔

AaronMoat replied:
Looks like @rarkins was thinking similarly with the code suggestion below, however I'm fairly confident it's always at minimum an empty array:

- https://github.com/renovatebot/renovate/blob/09743429d18b7778a95259de2b67fa3d83c45713/lib/util/git/index.ts#L232
- https://github.com/renovatebot/renovate/blob/09743429d18b7778a95259de2b67fa3d83c45713/lib/util/git/index.ts#L323
- https://github.com/renovatebot/renovate/blob/09743429d18b7778a95259de2b67fa3d83c45713/lib/util/git/types.ts#L33

rarkins commented on lib/util/git/index.ts:
```diff
@@ -673,19 +673,44 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+    includedAuthors.delete(gitAuthorEmail);
+  }
+
+  for (const ignoredAuthor of ignoredAuthors) {
```
```suggestion
  for (const ignoredAuthor of ignoredAuthors ?? []) {
```

rarkins reviewed (commented):
Does this make or require any change to caching?

AaronMoat commented:
Hmm, good question on caching changes. 

One potential thing is that the cache is solely based on the branch name and not the base branch. I guess if you now call isBranchModified for multiple different base branches then you could get into trouble? Is there a use case for this? I'll take your guidance on if we need to fix or not.

rarkins commented on lib/util/git/index.ts:
```diff
@@ -673,19 +673,44 @@ export async function isBranchModified(branchName: string): Promise<boolean> {
+
+  if (includedAuthors.size === 0) {
+    // authors all match - branch has not been modified
+    logger.debug(
+      {
+        branchName,
+        baseBranch,
+        committedAuthors: [...committedAuthors],
+        includedAuthors: [...includedAuthors],
+        gitAuthorEmail,
+        ignoredAuthors,
+      },
+      'branch.isModified() = false',
+    );
```
This is a pretty long debug message to print every time a branch is unmodified. I would prefer that the metadata is in a trace log, or that the debug is greatly shortened

rarkins commented:
If the cashing is based on branchName + commitSha then it might already work fine. If a branch SHA was "unmodified" on a previous run then it shouldn't change to modified even if the base branch has received commits. There's potentially an edge case where a modified one changes to unmodified due to creative git use but in that case they can request a rebase/retry
